### PR TITLE
LoRA and DoRA PEFT support for Fine-Tuning TimesFM

### PR DESCRIPTION
**Pull request by tanmayshishodia**
Thank you for this great project and providing open source inference code. 

### What does this PR do? 

- [x] A generic fine-tuning pipeline which supports 4 fine-tuning strategies:
    - [x] Full Fine-Tuning
    - [x] [Linear Probing ](https://arxiv.org/abs/2302.11939) [fine-tunes only the residual blocks and the embedding layer]
    - [x] [LoRA](https://arxiv.org/abs/2106.09685) [fine-tunes only a small number of parameters by decomposing the weight matrices into low-rank matrices, making it efficient in terms of memory and computational resources]
    - [x] [DoRA](https://arxiv.org/abs/2402.09353v4) [an extension of LoRA that decomposes the pre-trained weight into magnitude and direction components and  exploits LoRA for ‌directional adaptation improving learning capacity and stability without additional inference overhead; accepted in ICML 24]
- [x] Add testing framework [pytest] 

### Why is this PR needed?
- Primary motivation for this PR is to leverage LoRA to enable efficient training of multiple adapter weights on various tasks, domains, and time-series datasets while maintaining the same base model.
- Implementation of PEFT techniques are largely unexplored on time series foundational models. Help the research community analyze PEFT on these models.

### Performance Comparison of LoRA/DoRA with Linear Probing
image

Experiments were performed with a split of 60-20-20 train, val, and test split. Black denotes best, blue denotes second best.
Benchmarking was done with `context_len=128` and `horizon_len=96`, and fine-tuning was done with `context_len=128` and `horizon_len=128`.

### Caveats
- The loading of adapter weights is currently under optimized.
- I am new to the PaxML framework. Please assist with optimization wherever possible.

### Functional Testing
- [x] Test FFT all params get updated.
- [x] Test Linear Probing only residual block params get updated.
- [x] Test only LoRA params get updated in LoRA FT. Vary rank r.
- [x] Test LoRA and DoRA magnitude vector is updated in DoRA FT. Vary rank r.

### Ref PRs
- https://github.com/google/paxml/pull/83
- https://github.com/huggingface/peft/pull/1474

### CODE REVIEW

**Comment by rajatsen91** [comment]
Thanks @tanmayshishodia. These are very welcome contributions. Since the CL is pretty big it will take us some time to review and merge:

1. We added a notebook that can do linear probing (as well as full finetuning) under notebooks/covariates.ipynb. This really needs just a one line change. The rest of the notebook is just writing the training loop and dealing with paxml models. It might make sense to add LoRA and DoRA examples to that notebook.

2. A general style nit is that we are trying to not import layers individually but the whole module. For instance instead of `from praxis.layers.linears import Linear` we would prefer `from praxis.layers import linears` and then do linears.LInear. 

3. Please let us know when your changes are ready to review and @siriuz42 and I can take a shot at reviewing.

**Comment by rajatsen91** [comment]
4. I think on our side we need to also set up testing so that we can ask contributors to add tests to major changes like this one.

**Comment by tanmayshishodia** [comment]
Sure. 

1. This script is largely modified from the example notebook shared in `notebooks/finetuning.ipynb`. I can add an example of LoRA/DoRA there. The script shall be useful as it let's you run multiple experiments simultaneously with different configurations. I will be adding `.sh` for common configurations to quickly run them, [for example](https://github.com/google-research/timesfm/blob/a59979d3b429671bbb4a5fafb4c7ff8041e214ca/peft/dora.sh).
2. Sure, will make changes to address this.
3. Sure. I will change PR status to ready for review. There are some issues/bottlenecks I am aware of, I will mention them so you can help address them.
4. I can set it up with `pytest`, if that sounds good? Shall I keep it in a separate PR or this one?

Feel free to ask me any other questions you may have.

**Comment by rajatsen91** [comment]
SGTM to all. For 4, pytest sounds fine.

You can add some tests to this PR it self. Something that covers the __call__ function of the layers should be fine. Let us know if you have any praxis or paxml related questions.

**Comment by tanmayshishodia** [comment]
1. The current workflow for loading the model for adapter fine-tuning is:
    - Load the base model checkpoint which loads `train_state` and `jit_decodes` the model [`tfm`].
    - Create an instance of finetune model [`model`].
    - Replace the attention and linear layers in the`stacked transformer block` of the `model` with LoRA/DoRA layers defined in `adapters/lora_layers.py`, `adapters/dora_layers.py`. [[ref](https://github.com/google-research/timesfm/blob/a59979d3b429671bbb4a5fafb4c7ff8041e214ca/src/adapter/utils.py#L279)]
    - Since `tfm` was loaded with base model checkpoint, I have to manually add the LoRA/DoRA adapter weights for each layer. [[ref](https://github.com/google-research/timesfm/blob/a59979d3b429671bbb4a5fafb4c7ff8041e214ca/src/adapter/utils.py#L352)]. The setup method defined in the layer files don't aid in performing initialization. 

I believe there should be a better way to load the checkpoint along with the initialized lora and dora weights. I did try replacing the layers, instantiating and then loading the checkpoint but could not do so successfully. Let me know your thoughts on this.

2. After fine-tuning only the adapter weights are saved which I manually extract using [ref](https://github.com/google-research/timesfm/blob/a59979d3b429671bbb4a5fafb4c7ff8041e214ca/src/adapter/utils.py#L36). The loading of the fine-tuned model is as follows:
    - Load the base model ckpt and jit_decode.
    - To load the adapter ckpt I first create the config for a adapter model and extract the necessary `var_weight_hparams` to load the adapter ckpt and then merge them using the same forward pass logic and then again jit_decode the model.

This whole process can be more optimized I believe. Let me know how that can be done and if you have any ideas. 

3. Although the initialization of the adapter weights is not being done using setup method currently, `dora_m` param is initialized with the column norm of the pre-trained weight. How to do so with `WeightInit` in setup method so that it can be used later?

**Comment by rajatsen91** [comment]
Hi @tanmayshishodia,

Regarding "Replace the attention and linear layers in the stacked transformer block of the model with LoRA/DoRA layers defined in adapters/lora_layers.py, adapters/dora_layers.py"

afaik LoRA adds a low rank adapter additively to the original weights which are held fixed as in $Wx + \Delta W x$. However, here you are replacing the original attention weights. May be this is just a terminology issue and you are just adding $\Delta W$ and not removing $W$ ?

**Comment by tanmayshishodia** [comment]
Hi @rajatsen91 

Apologies for framing it incorrectly. The LoRA/DoRA layers defined in the files inherit original Linear and attention layers. While doing the forward pass we get the original weight matrix and add the LoRA delta (A and B which multiply to form delta W) while the original weight matrix is fixed. [[ref](https://github.com/tanmayshishodia/timesfm/blob/5901805e7efb2be70cdf8d25a984d16353711536/src/adapter/lora_layers.py#L52)].

sharing a diff between weights before training LoRA/DoRA and after a training epoch: https://www.diffchecker.com/SOotQLkx/.

**Comment by tanmayshishodia** [comment]
@rajatsen91 @siriuz42 PR is ready for first round of review. I have added a simple test case which tests inference. I will be adding more, let me know what scenarios need to be covered.

**Comment by rghosh08** [comment]
@tanmayshishodia We, at Nutanix, greatly appreciate your effort as an intern in fine-tuning TimesFM. As your mentor, I am super proud to see this. 

@rajatsen91 We, at Nutanix, are looking forward to contributing in your TimesFM project. We believe this will have significant impact across the industries. Thanks!

**Comment by rajatsen91** [comment]
Thanks again for the PR. Our team is traveling this week, will look into this next week

File: peft/fft.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
```
Comment (rajatsen91) [inline peft/fft.sh:2]: I feel like different .sh scripts are not needed, one script with command line options is good enough. or we can not check in these scripts and have it as example usage in the header comment of finetune.py

**Review by rajatsen91 — COMMENTED** [review]
LGTM overall. Left a minor comment about the shell scripts.

**Comment by rajatsen91** [comment]
The numbers in the table "Performance Comparison of LoRA/DoRA with Linear Probing" does not match the numbers I get from the finetuning.ipynb notebook that I had. Can you please clarify what are the differences ?

In particular for ETTm1 test split I get MAE: 0.351 for the base model.

File: peft/linear_probing.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    --context-len=512 \
+    --freq="15min" \
+    --data-path="../datasets/ETT-small/ETTm1.csv" \
+    --num-epochs=100 \
+    --learning-rate=1e-2 \
+    --adam-epsilon=1e-7 \
+    --adam-clip-threshold=1e2 \
+    --early-stop-patience=10 \
+    --datetime-col="date" \
+    --boundaries=1000 46080 57600 \
```
Comment (rajatsen91) [inline peft/linear_probing.sh:18]: why is this 1000 here?

Comment (tanmayshishodia) [reply peft/linear_probing.sh:18]: My bad, thank you for pointing it out. I will remove this param in the .sh file. The script automatically takes the split as 60-20-20.

**Comment by tanmayshishodia** [comment]
> The numbers in the table "Performance Comparison of LoRA/DoRA with Linear Probing" does not match the numbers I get from the finetuning.ipynb notebook that I had. Can you please clarify what are the differences ?
> 
> In particular for ETTm1 test split I get MAE: 0.351 for the base model.

Hi Rajat, all my experiments were run with boundaries of 60-20-20 % of the given dataset. I just noticed that even though the boundaries in notebook are spaced in the same configuration they aren't using the whole dataset. for ettm1 number of datapoints is 69680, however test boundary is till 57600 in the notebook. Hence it does not match. Can you try running the notebook with boundaries: `[41808, 55744, 69680]` for ettm1, you should get the same result.

File: peft/fft.sh (hunk 1 above)
Comment (tanmayshishodia) [reply peft/fft.sh:2]: Okay, I will keep only one of them. 

Running `python3 finetune.py --help` gives the following o/p, which is self-explanatory. Perhaps I can add it in README?
image

**Comment by rajatsen91** [comment]
Ok sounds great. I think from my side it is ready to merge, great work. It would be great if you can add a README.md in the peft folder along with the results in your table. Thanks again for the great contribution.

**Comment by rajatsen91** [comment]
LGTM from my side. Will wait to hear from  @siriuz42 and try to merge by EOD.

**Pull request merged**